### DISCUSSION TITLE
after crop page,  then get_images()

### DISCUSSION THREAD

**Discussion by yonglee7015** [pymupdf/PyMuPDF#2140] [Q&A] [answered]
HI  I just want to get_images in specify range in a page. So I crop the page, then use function get_images. But it still get all images in this page.

page.set_cropbox(fitz.Rect(100, 100, 550, 700))
image_list = page.get_images(full=True)

How to get images in specify range in a page?

**Comment by JorjMcKie** [comment, marked as answer]
`page.get_images()` may even list images that are **_not at all used by the page_** - not to mention what you are asking.

This is because that method looks at PDF definitions only and does not inspect the page's appearance source code.

To at least restrict the above list to images that are actually referenced by this page **_anywhere on the MediaBox_**, do a `page.clean_contents()` first.

To restrict that list to **_visible_** images (CropBox) do this:
```python
imglist = page.get_images()
visibles = [item for item in imglist if page.get_image_rects(item[0])[0] in page.cropbox]
```
`get_image_rects` walks through the page's appearance instructions to determine each bbox of one image (given by its xref `item[0]`) on the page. In the above I am looking at the first such bbox only to see whether it is in the CropBox.
A more precise code would have to go over the potentially several such rectangles - and may even include images that are at least partially visible by checking non-empty rectangle intersections.

**Reply by yonglee7015** [reply]
Thank you so much.  But it seems not as I want. I test it in my code

![image](https://user-images.githubusercontent.com/59268746/208386209-ca080e4a-9c9c-42ef-b4e3-8b75d2efc3b6.png)
![image](https://user-images.githubusercontent.com/59268746/208386284-13ddebed-40c9-45a1-bf49-3a625c0aa72a.png)

**Reply by yonglee7015** [reply]
Anyway  I just want to extract images in  a fixed position in pdf. based on rect can't get the correct image I need.

**Comment by JorjMcKie** [comment]
There must be some misconception. Can you let me have the PDF and the page number in question?

**Reply by yonglee7015** [reply]
Ok,just sent pdf to your email

**Comment by JorjMcKie** [comment]
As I suspected, you have been hiding major information items!
Your pages are rotated by 90°! As documented, **_every_** coordinates read or written however are (resp. must be) in **_unrotated_** coordinates.
The PDF `CropBox` itself **_always_** is in unrotated coordinates (in contrast to `page.rect`), so for setting this you need unrotated coords.
So if you extract image bboxes on a page, they will appear as if the page had not been rotated. This is also the coordinates you need for CropBox setting. If you do that and use this `Rect(109.0, 241.76995849609375, 529.0, 550.219970703125)` things will work fine.

**Reply by yonglee7015** [reply]
Haha, thank you. Actually I noticed rotation is 90. I'm not very familiar with a lot of the parameters in it.So I don't know how to fix it. Thanks again.

**Reply by JorjMcKie** [reply]
> Haha, thank you. Actually I noticed rotation is 90. I'm not very familiar with a lot of the parameters in it.So I don't know how to fix it. Thanks again.

But now you are on top of things I hope?
Admittedly, the coordinate system change between PDF <> (Py)MuPDF <> rotation-yes/no is easy to get confused with. Used to be the same with me.
After several unsatisfying alternatives, I finally decided to use the current solution - as described. You always will see (and have to specify) unrotated coordinates, underlying code will map it to the correct page location depending on the currently chosen rotation.

**Reply by yonglee7015** [reply]
it works for the pdf I sent to you. But this method is not flexible.  it not works for my other sample pdf files. Image in other pdfs returns different rect parameters, it is dynamic. so it would't fall in the rect parameters we predefined. I need to get images from a batch of pdf files.

**Reply by yonglee7015** [reply]
I found the rect of image at the same position in pdf  returns different rect parameters. so it is hard to find the image what I indeed to find.

how dos Pymudf handle the rect of pdf?  rect seems not like a location as document said,  x0,y0,x1,y1 more like a size of image.

**Reply by JorjMcKie** [reply]
> how dos Pymudf handle the rect of pdf? rect seems not like a location as document said, x0,y0,x1,y1 more like a size of image.

PyMuPDF of course uses MuPDF's algorithm. And that works like this (in PDF - for other document types there are deviating ones):
* If a page has a `/CropBox` key, it is used, but with the origin positioning top-left always at (0, 0) and always respecting any rotation.
* If a `/CropBox` does not exist, the `/MediaBox` is taken instead.
* If `/MediaBox` does not exist either, the Letter format is assumed.

So this is strictly following what PDF specifications are telling us and has nothing to with any images.

> I found the rect of image at the same position in pdf returns different rect parameters. so it is hard to find the image what I indeed to find

This is hard to believe. It sounds more like the algorithm to locate the desired page positions is still inadequate - or may be impossible to develop because of your complex situation.

**Reply by JorjMcKie** [reply]
Maybe you can use `Page` matrices to compute image positions relative to page rotation.
There are `page.rotation_matrix` and `page.derotation_matrix`. If you get an image's position as `bbox` coordinates of the unrotated page, then `bbox * page.rotation_matrix` is the image position on the rotated page, and so on.

**Reply by yonglee7015** [reply]
HI Jorj X. McKie, thanks for you help. Finally I found pdfplumber can solve my problem. Cos pdfplumber can export crop page to image directly.  So I set a fixed crop bbox in the page,
Then it can always export the image I want. 


!apt install imagemagick

import pdfplumber

pdf = pdfplumber.open('/content/015-23C110164.pdf')
currentpage = pdf.pages[0]
currentpage_crop = currentpage.crop((120,130,700,500))
image_obj = currentpage_crop.to_image(resolution=300)
image_obj.save('dddd.png', format="PNG")

**Reply by JorjMcKie** [reply]
Ah, it is only now that I understand what you actually wanted!

With the current solution you do not actually **_extract_** the underlying image of the PDF:
You are simply taking a "photo" of a certain page area.
That would have been even easier in PyMuPDF, if we had been communicating better:
```python
pix = page.get_pixmap(dpi=300, clip=(120,130,700,500))
pix.save("dddd.png")
```

**Reply by yonglee7015** [reply]
Sorry for my poor English. I can't clearly express my meaning. Image rect is so dynamics,this way should be the best.

**Reply by JorjMcKie** [reply]
That's ok. BTW I am sure that PyMuPDF is faster here, too.

What brought me on the wrong track was that talk about changing the cropbox. But this was not at all what you wanted - you needed a subimage of the full page like a photo.

**Reply by yonglee7015** [reply]
Thanks. It is faster and simple.   I need to read the pymupdf document carefully first.
